src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -171,6 +202,41 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
 
     } while (next != ROPE_SENTINEL);
   }
+
+  template <typename T = Callback>
+  KOKKOS_FUNCTION std::enable_if_t<
+      std::is_same<T, Callback>{} && !traverse_half<Callback>{}, bool>
+  pruneLeaf(int, int) const
+  {
+    return false;
+  }
+
+  template <typename T = Callback>
+  KOKKOS_FUNCTION
+      std::enable_if_t<std::is_same<T, Callback>{} && traverse_half<Callback>{},
+                       bool>
+      pruneLeaf(int query_index, int leaf_index) const
+  {
+    int const leaf_nodes_shift = _bvh.size() - 1;
+    return leaf_index - leaf_nodes_shift <= query_index;
+  }
+
+  template <typename T = Callback>
+  KOKKOS_FUNCTION std::enable_if_t<
+      std::is_same<T, Callback>{} && !traverse_half<Callback>{}, bool>
+  pruneLeftSubtree(int, int) const
+  {
+    return false;
+  }
+
+  template <typename T = Callback>
+  KOKKOS_FUNCTION
+      std::enable_if_t<std::is_same<T, Callback>{} && traverse_half<Callback>{},
+                       bool>
+      pruneLeftSubtree(int query_index, int left_child_index) const
+  {
+    return left_child_index <= query_index;
```

Comment:
So, we need to think a bit further about what we are comparing here.

First of all, we don't know if the predicates are sorted or not. It would work with both, but the meaning is going to be different.

Second, lets say the predicates *are* sorted. Then the `queryIndex` matches the leaf node indices, so distant leaves should have distant indices, and not overlap. I think that's OK. The way the algorithm prunes then depends on the spatial curve ordering. So, then, one cannot assume that geometrically you would only get primitives from one side of a sphere, for example.

I wonder, though, what would happen if we tried to replace `pruneLeaf` with comparing stored permutation with queryIndex? One thing it would certainly achieve is that the resulting indices would be on one side of the `queryIndex`. Not sure what it means, though, but would like to try and see the results.

Reply:
After thinking a bit about it, there's not much we can do here. Essentially, our answer is going to be "Find all pairs (i,j) satisfying both predicate and morton_codes(i) <= morton_codes(j)". Changing the latter condition to "i < j" is not possible, as otherwise subtrees cannot be skipped.

This actually made me realize that our output in the crs graph is always in the order of increasing Morton indices, and not in the increasing order of indices. This could actually cause some slowdown in the force calculation even when a regular ArborX calls are used.